bench/replay.ml

```diff
@@ -0,0 +1,266 @@
+
+module Trace = struct
+  type key = string [@@deriving repr]
+
+  type op =
+    | Clear
+    | Flush
+    | Mem of key * bool
+    | Find of key * bool
+    | Ro_mem of key * bool
+    | Ro_find of key * bool
+    | Add of key * (int63 * int * char)
+  [@@deriving repr]
+
+  let open_ops_sequence path : op Seq.t =
+    let chan = open_in_bin path in
+    decoded_seq_of_encoded_chan_with_prefixes op_t chan
+end
+
+module Benchmark = struct
+  type result = { time : float; size : int }
+
+  let run config f =
+    let time, res = with_timer f in
+    let size = FSHelper.get_size config.root in
+    ({ time; size }, res)
+
+  let get_maxrss () =
+    let usage = Rusage.(get Self) in
+    let ( / ) = Int64.div in
+    usage.maxrss / 1024L / 1024L
+
+  let pp_results ppf result =
+    Format.fprintf ppf "Total time: %f; Size on disk: %d M; Maxrss: %Ld"
+      result.time result.size (get_maxrss ())
+end
+
+module type S = sig
+  include Index.S
+
+  val v : string -> t
+
+  val close : t -> unit
+end
+
+module Index = struct
+  module Index =
+    Index_unix.Make (Encoding.Key) (Encoding.Val) (Index.Cache.Unbounded)
+  include Index
+
+  let cache = Index.empty_cache ()
+
+  let v root =
+    Index.v ~cache ~readonly:false ~fresh:false ~log_size:500_000 root
+
+  let close t = Index.close t
+end
+
+module Bench_suite
+    (Store : S
+               with type key = Encoding.Hash.t
+                and type value = int63 * int * char) =
+struct
+  let key_to_hash k =
+    match Encoding.Hash.of_string k with
+    | Ok k -> k
+    | Error (`Msg m) -> Fmt.failwith "error decoding hash %s" m
+
+  let add_operation store op_seq nb_ops () =
+    with_progress_bar ~message:"Replaying trace" ~n:nb_ops ~unit:"operations"
+    @@ fun progress ->
+    let rec aux op_seq i =
+      if i >= nb_ops then i
+      else
+        match op_seq () with
+        | Seq.Nil -> i
+        | Cons (op, op_seq) ->
+            let () =
+              match op with
+              | Trace.Flush -> Store.flush store
+              | Clear -> Store.clear store
+              | Mem (k, b) ->
+                  let k = key_to_hash k in
+                  let b' = Store.mem store k in
+                  if b <> b' then
+                    Fmt.failwith "Operation mem %a expected %b got %b"
+                      (Repr.pp Encoding.Key.t) k b b'
+              | Find (k, b) ->
+                  let k = key_to_hash k in
+                  let b' =
+                    try
+                      let _ = Store.find store k in
+                      true
+                    with Not_found -> false
+                  in
+                  if b <> b' then
+                    Fmt.failwith "Operation find %a expected %b got %b"
+                      (Repr.pp Encoding.Key.t) k b b'
+              | Add (k, v) ->
+                  let k = key_to_hash k in
+                  Store.replace store k v
+              | Ro_mem _ | Ro_find _ -> ()
+            in
+            progress Int64.one;
```

Comment:
This is going to be rendered a lot. It might currently slow down the replay.

Try adding a `~sampling_interval` to `Progress_unix.counter`. Aim for 1 to 100 renders per sec.

Reply:
(Aside: I will fix this tediousness in the next release of Progress...)